sys/gen/pmap.c

```diff
@@ -0,0 +1,593 @@
+  SCOPED_MTX_LOCK(&user_pmaps_lock);
+
+  pmap_t *user_pmap;
+  LIST_FOREACH (user_pmap, &user_pmaps, pmap_link) {
+    pde_t *pdep = (pde_t *)phys_to_dmap(user_pmap->pde);
+    pdep[idx] = pde;
+  }
+}
+#endif /* !SHARED_KERNEL_PD */
+
+/* Return PTE pointer for `va`. Allocate page table if needed. */
+static pte_t *pmap_ensure_pte(pmap_t *pmap, vaddr_t va) {
+  assert(mtx_owned(&pmap->mtx));
+
+  pde_t *pdep = (pde_t *)phys_to_dmap(pmap->pde) + pt_index(0, va);
+
+  for (unsigned lvl = 0; lvl < PAGE_TABLE_DEPTH - 1; lvl++) {
+    paddr_t pa;
+    if (!pde_valid_p(*pdep)) {
+      pa = pmap_alloc_pde(pmap, va);
+      *pdep = pde_make(lvl, pa);
+#if SHARED_KERNEL_PD
+      if (lvl == 0 && pmap == pmap_kernel())
+        broadcast_kernel_top_pde(pt_index(0, va), *pdep);
+#endif /* !SHARED_KERNEL_PD */
+    } else {
+      pa = pte_frame((pte_t)*pdep);
+    }
+    pdep = (pde_t *)phys_to_dmap(pa) + pt_index(lvl + 1, va);
+  }
+
+  return (pte_t *)pdep;
+}
+
+/*
+ * Wired memory interface.
+ */
+
+void pmap_kenter(vaddr_t va, paddr_t pa, vm_prot_t prot, unsigned flags) {
+  pmap_t *pmap = pmap_kernel();
+
+  assert(page_aligned_p(pa) && page_aligned_p(va));
+  assert(pmap_address_p(pmap, va));
+
+  klog("Enter unmanaged mapping from %p to %p", va, pa);
+
+  pte_t pte = pte_make(pa, prot, flags, true);
+
+  WITH_MTX_LOCK (&pmap->mtx) {
+    pte_t *ptep = pmap_ensure_pte(pmap, va);
+    pmap_write_pte(pmap, ptep, pte, va);
+  }
+}
+
+void pmap_kremove(vaddr_t va, size_t size) {
+  pmap_t *pmap = pmap_kernel();
+
+  assert(page_aligned_p(va) && page_aligned_p(size));
+  assert(pmap_contains_p(pmap, va, va + size));
+
+  klog("Remove unmanaged mapping for %p - %p range", va, va + size - 1);
+
+  WITH_MTX_LOCK (&pmap->mtx) {
+    for (size_t off = 0; off < size; off += PAGESIZE) {
+      pte_t *ptep = pmap_lookup_pte(pmap, va + off);
+      assert(ptep);
+      pmap_write_pte(pmap, ptep, pte_empty(true), va + off);
+    }
+  }
+}
+
+bool pmap_kextract(vaddr_t va, paddr_t *pap) {
+  return pmap_extract(pmap_kernel(), va, pap);
+}
+
+/*
+ * Pageable user memory interface.
+ */
+
+void pmap_enter(pmap_t *pmap, vaddr_t va, vm_page_t *pg, vm_prot_t prot,
+                unsigned flags) {
+  assert(pmap != pmap_kernel());
+
+  paddr_t pa = pg->paddr;
+
+  assert(page_aligned_p(va));
+  assert(pmap_address_p(pmap, va));
+
+  klog("Enter virtual mapping %p for frame %p", va, pa);
+
+  pte_t pte = pte_make(pa, prot, flags, false);
+
+  WITH_MTX_LOCK (&pv_list_lock) {
+    WITH_MTX_LOCK (&pmap->mtx) {
+      pv_entry_t *pv = pv_find(pmap, va, pg);
+      if (!pv)
+        pv_add(pmap, va, pg);
+      pg->flags &= ~(PG_MODIFIED | PG_REFERENCED);
+      pte_t *ptep = pmap_ensure_pte(pmap, va);
+      pmap_write_pte(pmap, ptep, pte, va);
+    }
+  }
+}
+
+void pmap_remove(pmap_t *pmap, vaddr_t start, vaddr_t end) {
+  assert(pmap != pmap_kernel());
+  assert(page_aligned_p(start) && page_aligned_p(end) && start < end);
```

Comment:
For me the last part of this check should be done in `pmap_contains_p`.

Reply:
Done.